File: sonar-text-plugin/src/test/java/org/sonar/plugins/secrets/rules/AlibabaCloudAccessKeySecretsRuleTest.java
/*
 * SonarQube Text Plugin
 * Copyright (C) 2021-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plugins.secrets.rules;

import java.util.List;
import org.junit.Test;
import org.sonar.plugins.secrets.TestUtils;

import static java.nio.charset.StandardCharsets.UTF_8;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.groups.Tuple.tuple;
import static org.sonar.plugins.secrets.TestUtils.aNormalizedInputFile;

public class AlibabaCloudAccessKeySecretsRuleTest {

  AlibabaCloudAccessKeySecretsRule underTest = new AlibabaCloudAccessKeySecretsRule();

  @Test
  public void testRuleProperties() {
    assertThat(underTest.getRuleKey()).isEqualTo("S6336");
    assertThat(underTest.getMessage()).isEqualTo("Make sure this Alibaba Cloud Access Key Secret is not disclosed.");
  }

  @Test
  public void testRuleRegex1Positive() {
    List<Secret> secrets = underTest.findSecretsIn(TestUtils.aNormalizedInputFile("String aliyunAccessKeySecret=\"KmkwlDrPBC68bgvZiNtrjonKIYmVT8\";"));

    assertThat(secrets)
      .extracting(Secret::getTextRange)
      .extracting("start.line", "start.lineOffset", "end.line", "end.lineOffset")
      .containsOnly(tuple(1, 30, 1, 60));
  }

  @Test
  public void testRuleRegex2Positive() {
    List<Secret> secrets = underTest.findSecretsIn(TestUtils.aNormalizedInputFile("static string AccessKeySecret = \"l0GdwcDYdJwB1VJ5pv0ormyTV9nhvW \";"));

    assertThat(secrets)
      .extracting(Secret::getTextRange)
      .extracting("start.line", "start.lineOffset", "end.line", "end.lineOffset")
      .containsOnly(tuple(1, 33, 1, 63));
  }

  @Test
  public void testRuleRegexNegative() throws Exception {
    List<Secret> secrets = underTest.findSecretsIn(TestUtils.readFileAndNormalize("src/test/files/google-cloud-account-key/GoogleCloudAccountNegative.json", UTF_8));

    assertThat(secrets).isEmpty();
  }

  @Test
  public void testRuleRegexExamplePositive() {
    List<Secret> secrets = underTest.findSecretsIn(TestUtils.aNormalizedInputFile("String aliyunAccessKeySecret=\"KmkwlDrPBC68bgvZiNtrjonKIYmVT8\";"));

    assertThat(secrets)
      .extracting(Secret::getTextRange)
      .extracting("start.line", "start.lineOffset", "end.line", "end.lineOffset")
      .containsOnly(tuple(1, 30, 1, 60));
  }

  @Test
  public void testRuleRegexExampleNegativeLowEntropy() {
    List<Secret> secrets = underTest.findSecretsIn(TestUtils.aNormalizedInputFile("String aliyunAccessKeySecret=\"100000000000000000000000000000\";"));

    assertThat(secrets).isEmpty();
  }

}

File: sonar-text-dotnet/src/SonarLint.Secrets.DotNet.UnitTests/Rules/AlibabaCloudAccessKeySecretsRuleTest.cs
/*
 * Copyright (C) 2021-2023 SonarSource SA
 * All rights reserved
 * mailto:info AT sonarsource DOT com
 */

// Ported from ...\sonar-secrets-plugin\src\test\java\com\sonarsource\secrets\rules\AlibabaCloudAccessKeySecretsRuleTest.java

using FluentAssertions;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using SonarLint.Secrets.DotNet.Rules;
using SonarLint.VisualStudio.Integration.UnitTests;
using System.Linq;
using static SonarLint.Secrets.DotNet.UnitTests.JavaTestFileUtils;
using static SonarLint.Secrets.DotNet.UnitTests.TestUtils;

namespace SonarLint.Secrets.DotNet.UnitTests.Rules
{
    [TestClass]
    public class AlibabaCloudAccessKeySecretsRuleTest
    {
        [TestMethod]
        public void MefCtor_CheckIsExported()
        {
            MefTestHelpers.CheckTypeCanBeImported<AlibabaCloudAccessKeySecretsRule, ISecretDetector>(null, null);
        }

        [TestMethod]
        public void TestRuleProperties()
        {
            var testSubject = new AlibabaCloudAccessKeySecretsRule();

            testSubject.RuleKey.Should().Be("secrets:S6336");
            testSubject.Name.Should().Be("Alibaba Cloud AccessKeys should not be disclosed");
            testSubject.Message.Should().Be("Make sure this Alibaba Cloud Access Key Secret is not disclosed.");
        }

        [TestMethod]
        public void testRuleRegex1Positive()
        {
            var testSubject = new AlibabaCloudAccessKeySecretsRule();

            var input = "String aliyunAccessKeySecret=\"KmkwlDrPBC68bgvZiNtrjonKIYmVT8\";";

            var secrets = testSubject.Find(input);

            secrets.Count().Should().Be(1);
            CheckExpectedSecretFound(input, "KmkwlDrPBC68bgvZiNtrjonKIYmVT8", secrets.First());
            CrossCheckWithJavaResult(input, new(1, 30, 1, 60), secrets.First());
        }

        [TestMethod]
        public void testRuleRegex2Positive()
        {
            var testSubject = new AlibabaCloudAccessKeySecretsRule();

            var input = "static string AccessKeySecret = \"l0GdwcDYdJwB1VJ5pv0ormyTV9nhvW \";";

            var secrets = testSubject.Find(input);

            secrets.Count().Should().Be(1);
            CheckExpectedSecretFound(input, "l0GdwcDYdJwB1VJ5pv0ormyTV9nhvW", secrets.First());
            CrossCheckWithJavaResult(input, new(1, 33, 1, 63), secrets.First());
        }

        [TestMethod]
        public void TestRuleRegexNegative()
        {
            var testSubject = new AlibabaCloudAccessKeySecretsRule();

            var secrets = testSubject.Find(readFileAndNormalize(TestConstants.RootTestFileDirectory + "checks\\GoogleCloudAccountKeyCheck\\GoogleCloudAccountNegative.json", UTF_8));

            secrets.Should().BeEmpty();
        }

        [TestMethod]
        public void testRuleRegexExamplePositive()
        {
            var testSubject = new AlibabaCloudAccessKeySecretsRule();

            var input = "String aliyunAccessKeySecret=\"KmkwlDrPBC68bgvZiNtrjonKIYmVT8\";";

            var secrets = testSubject.Find(input);

            secrets.Count().Should().Be(1);
            CheckExpectedSecretFound(input, "KmkwlDrPBC68bgvZiNtrjonKIYmVT8", secrets.First());
            CrossCheckWithJavaResult(input, new(1, 30, 1, 60), secrets.First());
        }

        [TestMethod]
        public void testRuleRegexExampleNegativeLowEntropy()
        {
            var testSubject = new AlibabaCloudAccessKeySecretsRule();

            var secrets = testSubject.Find("String aliyunAccessKeySecret=\"100000000000000000000000000000\";");

            secrets.Should().BeEmpty();
        }
    }
}
